core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file;
+
+import alluxio.master.file.meta.MutableInodeDirectory;
+import alluxio.master.file.meta.MutableInodeFile;
+import alluxio.master.journal.JournalEntryMerger;
+import alluxio.proto.journal.Journal;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A file system journal entry merger which merges inode creation and inode update journals
+ * on the same inode object into one.
+ * This class is not thread-safe and should not be shared across threads.
+ */
+@NotThreadSafe
+public class FileSystemJournalEntryMerger implements JournalEntryMerger {
+  /** Persists merged journal entries. */
+  private final List<Journal.JournalEntry> mJournalEntries = new ArrayList<>();
+  /** A map whose key is the inode id and value is the journal index in the mJournalEntries list.*/
+  private final Map<Long, Integer> mEntriesMap = new HashMap<>();
+
+  /**
+   * Adds a new journal entry and merges it on the fly.
+   * @param entry the new journal entry to add
+   */
+  @Override
+  public void add(Journal.JournalEntry entry) {
```

Comment:
We are only merging Update into the inodes here right? Any plans for other merges like merging Create and Delete? I guess the intention here is that Create and Update tend to be separate entries due to how those operations are implemented, so that's the only case we care about here?

Reply:
yeah i don't think we have any file system operation that creates a file and then deletes it. So probably this is okay and we don't have to merge them.